src/main/java/org/spongepowered/api/event/lifecycle/RegisterRegistryValueEvent.java

```diff
@@ -24,18 +24,18 @@
  */
 package org.spongepowered.api.event.lifecycle;
 
-import org.spongepowered.api.CatalogType;
+import org.spongepowered.api.ResourceKey;
 import org.spongepowered.api.event.GenericEvent;
 import org.spongepowered.api.registry.DuplicateRegistrationException;
+import org.spongepowered.api.registry.RegistryEntry;
+import org.spongepowered.api.registry.RegistryKey;
+import org.spongepowered.api.registry.ScopedRegistryHolder;
 
-public interface RegisterCatalogEvent<T extends CatalogType> extends GenericEvent<T>, LifecycleEvent {
+public interface RegisterRegistryValueEvent<H extends ScopedRegistryHolder, T> extends GenericEvent<T>, LifecycleEvent {
```

Comment:
Are we assuming that every registry holder can only have one registry of type `T`?

There is also going to need to be another form of `GenericEvent` that accepts two generic parameters - or we turn this into three events, one for each scope.

Reply:
No, in fact it is entirely possible for a `RegistryHolder` to have multiple registries of type `T`. This is why this event gives you the `RegistryKey` of the registry being invoked. Unfortunately, plugin devs will need to check against the API `Registries` constants if they want to register against a built-in registry. I don't see any clean away around it.

Reply:
No, neither do I. The only thing I can think of is adding a `Optional<RegistryEntry<T>> registerIf(RegistryKey<T> registryKey, ResourceKey key, T value)` if we want to allow plugin devs to be lazy about it - or perhaps make it the only register method to enforce devs to think about it.

There isn't a nice solution though, I agree.